### ISSUE TITLE
How can I display editorState outside the Editor component ? 

### ISSUE THREAD

**Issue by tony-go** [draft-js-plugins/draft-js-plugins#1098]
Hi Drafters :) 

Im trying to use DraftJs to have a rich editor for my Chat Input (hashtag and Emoji).
My Main problem is : **How can I display my messages after submit my Input to the feed** ? 

Should I re-used the Editor Component provide by draft JS ?
What is the best practice ?

Do you guys have some examples about chat rooms used DraftJS ? 

Thank you for

**Comment by freedomlang** [comment]
@TonyGorez wrap the editor as a component [with some config、plugins], then use that wraped component. btw here is the [slack](https://draftjs.slack.com) for draft js

**Comment by TonyGorez** [comment]
Thanks guys for your answers :) 

@efjacobson I thinked to user two differents `<Editor />` on for editing and an other one with the readOnly props(`<Editor _readOnly_/>`). 
Your idea is to pass all the messages in the same Editor component ?

**Comment by efjacobson** [comment]
@TonyGorez if i were going to do that, i would probably have a handleReturn (https://draftjs.org/docs/api-reference-editor.html) on the editing editor that took what had been entered and appended that to the contents of the readonly editor, then reset the editing editor to empty.

**Comment by tumetus** [comment]
Hello @TonyGorez !

I have worked on a slack-like chat room project using draft-js. My approach was pretty much what @efjacobson suggested.

My solution was something like this: I had two components `<Editor />` and` <ViewOnlyEditor />.`

`<Editor />` component wraps draft-js editor with plugins like emoji- and mention-plugin initialized. This component is used for inputting the messages.

`<ViewOnlyEditor />` component wraps draft-js editor with plugins like emoji-, linkify- and mention-plugin initialized and readOnly = true.

This allowed me to save the message from Editor-component as json (convertToRaw) to database. Then I could easily display the message by initialising the ViewOnlyEditor with the json from the database (convertFromRaw) and setting the readOnly = true.

So basically every message displayed was its own ViewOnlyEditor and then I had just one Editor at the bottom of the page that I could use to input new messages.

Hope this helps! Not sure if this is the best possible way to implement that kind of functionality but at least it worked well for my project. And if there is other solutions for this, I would be happy to hear them!

**Comment by juliankrispel** [comment]
@TonyGorez tbh although the draft js editor has a readonly flag there's a bunch of problems with it. From what I know from other folks implementing chat systems such as @mxstbr rendering your draft js editorState in a different component is the better approach. I'd recommend using something like [draft-convert](https://github.com/HubSpot/draft-convert) for rendering readonly draft js content ...

I'm going to label this with the documentation flag since I think it'd be useful to add it to the upcoming docs ...

**Comment by mxstbr** [comment]
We use `redraft` and it's served us very well on both ReactNative and the web! See https://github.com/lokiuz/redraft

**Comment by promerorgz** [comment]
@efjacobson 
> @TonyGorez if i understand you correctly, you would like to use draft as the input method for a chatroom, where once you submitted your message, it would disappear from draft and appear in the chatroom?
> 
> if that is correct, i would actually suggest not using draft to display your messages after submitting. draft is specifically designed to be an editor. depending on your use-case, on submitting a draft message, submit that data to some persistent storage (database if you want them to be permanent, redux or the like if it's ephemeral) and display it elsewhere.
> 
> for example, my app has a preview for what has been entered into draft. when the preview is clicked a modal pops up, grabs the editorState out of redux, and uses it to render what has been entered. the draft editor is specifically for editing and doesn't render the full preview, while the preview modal is totally fleshed out but does not allow editing (eg in draft a youtube embed block is an icon placeholder that can be edited, but in the preview modal it is the actual youtube embed iframe).
> 
> edit: i suppose if you were dead-set on using it to show the messages, you could have 2 separate instances of draft - one for editing and another for preview which is in read-only mode. on submit, you would refresh the read-only one with the new data from the editing one.

I am trying to do something similar to this.

> @TonyGorez if i understand you correctly, you would like to use draft as the input method for a chatroom, where once you submitted your message, it would disappear from draft and appear in the chatroom?
> 
> if that is correct, i would actually suggest not using draft to display your messages after submitting. draft is specifically designed to be an editor. depending on your use-case, on submitting a draft message, submit that data to some persistent storage (database if you want them to be permanent, redux or the like if it's ephemeral) and display it elsewhere.
> 
> for example, my app has a preview for what has been entered into draft. when the preview is clicked a modal pops up, grabs the editorState out of redux, and uses it to render what has been entered. the draft editor is specifically for editing and doesn't render the full preview, while the preview modal is totally fleshed out but does not allow editing (eg in draft a youtube embed block is an icon placeholder that can be edited, but in the preview modal it is the actual youtube embed iframe).
> 
> edit: i suppose if you were dead-set on using it to show the messages, you could have 2 separate instances of draft - one for editing and another for preview which is in read-only mode. on submit, you would refresh the read-only one with the new data from the editing one.

How do you display editorState in a non Editor component? Would you use the `convertFromRaw` function? I am trying to do something similar to this but I am having issues using two `Editor` components at the same time (one read only, one normal) so I agree that the Editor should just be for editing.

**Comment by efjacobson** [comment]
@promerorgz 

the way i use draft is specifically as an editor layer to an independent set of schemas. for text i take advantage of the native draft stuff, as well as several plugins and an inline entity layer for some custom stuff. for non-text (atomic) stuff, i actually set the entity's data payload to be an instance of something i want it to be. so at certain points (saving, previewing, etc) i convert the draftjs editorState into a format that is more useful for me.

for example, say that i want to use draft to make a youtube embed block. i have my own idea of what that should look like, for simplicity lets say it is just an object with `type`, `id`, `start_at`, and `autoplay`. when it is time to preview the content, i convert the _entire_ editorState into an array of these objects (by iterating through the draftjs ContentBlocks and extracting their entity data), and then in the preview component (which has no idea that draft exists) render them as i want to.

fwiw the conversion process runs reverse when i hydrate the editor. my persistent storage has no idea about draft, but on the front end i convert it into an editorState so i can use draft to edit it. once done editing, it gets converted back into my format and saved.

for converting from editorState -> my own stuff, i use https://github.com/sstur/draft-js-utils/tree/master/packages/draft-js-export-html as well my own custom code to fill in the blanks. last i checked, there are several similar utilities like this available, i recommend that you look into each of them to see which will work best for you. best of luck

**Comment by promerorgz** [comment]
@efjacobson thank you! This is very helpful!

**Comment by varunajmera0** [comment]
https://codeburst.io/web-developer-playbook-rich-text-editor-4c356fb8929d

you can use this. **Without any plugin**
